apps/meeting/src/providers/VideoGridProvider/state.tsx

```diff
@@ -0,0 +1,319 @@
+  VideoSource,
+} from 'amazon-chime-sdk-js';
+import { priorityBasedPolicy } from '../../meetingConfig';
+
+interface AttendeeStatus {
+  attendeeId: string;
+  name: string;
+  videoEnabled: boolean;
+  // sharingContent: boolean;
+}
+
+export type RosterAttendeeType = {
+  chimeAttendeeId: string;
+  externalUserId?: string;
+  name?: string;
+};
+
+export type RosterType = {
+  [attendeeId: string]: RosterAttendeeType;
+};
+
+export type Controls = {
+  zoomIn: () => void;
+  zoomOut: () => void;
+  prevPage: () => void;
+  nextPage: () => void;
+};
+
+export type State = {
+  // Keep the order of attendees based on Roster
+  attendees: { [attendeeId: string]: AttendeeStatus };
+  // Remote video sources sorted in the order of roster
+  availableVideoSources: string[];
+  // Pagination state
+  currentPage: number;
+  isZoomed: boolean;
+  zoomedLevel: number;
+  zoomedLevelIndex: number;
+};
+
+export enum VideoGridAction {
+  UpdateVideoSources,
+  ResetVideoGridState,
+  UpdateRoster,
+  ZoomIn,
+  ZoomOut,
+  PrevPage,
+  NextPage,
+}
+
+type UpdateVideoSources = {
+  type: VideoGridAction.UpdateVideoSources;
+  payload: {
+    videoSources: VideoSource[];
+  };
+};
+
+type ResetVideoGridState = {
+  type: VideoGridAction.ResetVideoGridState;
+  payload?: any;
+};
+
+type UpdateRoster = {
+  type: VideoGridAction.UpdateRoster;
+  payload: {
+    roster: RosterType;
+  };
+};
+
+type ZoomIn = {
+  type: VideoGridAction.ZoomIn;
+  payload?: any;
+};
+
+type ZoomOut = {
+  type: VideoGridAction.ZoomOut;
+  payload?: any;
+};
+
+type PreviousPage = {
+  type: VideoGridAction.PrevPage;
+  payload?: any;
+};
+
+type NextPage = {
+  type: VideoGridAction.NextPage;
+  payload?: any;
+};
+
+export type Action =
+  | UpdateVideoSources
+  | ResetVideoGridState
+  | UpdateRoster
+  | ZoomIn
+  | ZoomOut
+  | PreviousPage
+  | NextPage;
+
+export const initialState: State = {
+  attendees: {},
+  availableVideoSources: [],
+  currentPage: 1,
+  isZoomed: false,
+  zoomedLevel: 0,
+  zoomedLevelIndex: 0,
+};
+
+const zoomedLevels: number[] = [0, 1, 2, 4, 8, 16, 25];
+
+export function reducer(state: State, { type, payload }: Action): State {
+  const { attendees } = state;
+  let {
+    availableVideoSources,
+    currentPage,
+    isZoomed,
+    zoomedLevel,
+    zoomedLevelIndex,
+  } = state;
+
+  const updateZoomStatus = (): void => {
+    const size = availableVideoSources.length;
+    for (let i = 1; i < zoomedLevels.length - 1; i += 1) {
+      if (zoomedLevels[i] === size) {
+        zoomedLevelIndex = i;
+      } else if (zoomedLevels[i] < size) {
+        zoomedLevelIndex = i + 1;
+      }
+    }
+    zoomedLevel = zoomedLevels[zoomedLevelIndex];
+  };
+
+  const calculateVideoSourcesToBeRendered = (): string[] => {
+    const left = (currentPage - 1) * zoomedLevel;
+    const right = currentPage * zoomedLevel;
+    return availableVideoSources.slice(left, right);
+  };
+
+  const updateDownlinkPreferences = (): void => {
```

Comment:
Can we move these function out? This is too long to be nested function.

Reply:
Here it works as a closure, it's easier to access the `state`. If we move it out, we need to pass the states to these functions, which could make code worse. But I will try to see which one better.

Reply:
Moved these function out